File: src/SignUpBot.js
// Require selenium webdriver
let { Builder, Browser, By, Key, until } = require("selenium-webdriver");
let fs = require("fs");
let assert = require("assert");

async function click(driver, path) {
  await driver.findElement(By.xpath(path)).click();
}

async function signUpBot(account) {
  let driver = await new Builder().forBrowser("chrome").build();
  try {
    await driver.get(
      // replace with a link of your favorite sport booking URL
      "https://www.buchsys.ahs.tu-dortmund.de/angebote/aktueller_zeitraum/_Badminton.html"
    );

    //Store the ID of the original window
    const originalWindow = await driver.getWindowHandle();

    //Check we don't have other windows open already
    assert((await driver.getAllWindowHandles()).length === 1);

    // choose the right button according to the day of week
    // MUSS BE REPLACE WITH THE RIGHT XPATH IF YOU ENTER A NEW SPORT
    switch (new Date().getDay()) {
      //if monday
      case 1: {
        click(driver, '//*[@id="bs_trDFB6F6C0D3"]/td[9]/input');
        break;
      }
      //if tuesday
      case 2: {
        click(driver, '//*[@id="bs_trDFB6F6C0AC"]/td[9]/input');
        break;
      }
      //if thursday
      case 4: {
        click(driver, '//*[@id="bs_trDFB6F6C0E8"]/td[9]/input');
        break;
      }
      //if Friday
      case 5: {
        click(driver, '//*[@id="bs_trDFB6F60C0B"]/td[9]/input');
        break;
      }
      default: {
        await driver.quit();
        break;
      }
    }

    //Wait for the new window or tab
    await driver.wait(
      async () => (await driver.getAllWindowHandles()).length === 2,
      10000
    );

    //Loop through until we find a new window handle
    const windows = await driver.getAllWindowHandles();
    windows.forEach(async (handle) => {
      if (handle !== originalWindow) {
        await driver.switchTo().window(handle);
      }
    });

    //Wait for the new tab to finish loading content
    await driver.wait(until.titleIs("Anmeldung"), 5000);

    //chose booking with account
    await click(
      driver,
      '//*[@id="bs_form_main"]/div/div[2]/div[1]/label/div[2]/input'
    );

    const pathWithAccount = [
      // email input field
      '//*[@id="bs_pw_anm"]/div[2]/div[2]/input',
      //password input field
      '//*[@id="bs_pw_anm"]/div[3]/div[2]/input',
    ];

    click(driver, '//*[@id="bs_pw_anmlink"]/div[2]');
    // await driver
    //   .findElement(By.xpath('//*[@id="bs_pw_anmlink"]/div[2]'))
    //   .click();

    for (let i = 0; i < pathWithAccount.length; i++) {
      await driver
        .findElement(By.xpath(pathWithAccount[i]))
        .sendKeys(account[i]);
    }

    //submit form
    await click(driver, '//*[@id="bs_pw_anm"]/div[5]/div[1]/div[2]/input');

    // wait until all contents of the page is load
    await driver.wait(
      until.elementLocated(By.xpath('//*[@id="bs_bed"]/label/input'), 5000)
    );

    //click on agreement
    await click(driver, '//*[@id="bs_bed"]/label/input');

    //submit form
    await click(driver, '//*[@id="bs_submit"]');

    //book appointment
    await click(driver, '//*[@id="bs_foot"]/div[1]/div[2]/input');

    await driver.wait(until.titleIs("12"), 10000);
  } finally {
    await driver.quit();
  }
}

(async function book() {
  //list of accounts which do the booking
  const accounts = [
    // ["minhdan1405@gmail.com", "Minhdan1405@deutsch"],
    ["dangkhoa27031997@gmail.com", "B4byMet@l2703"],
  ];
  for (let account of accounts) {
    signUpBot(account);
  }
})();

File: signUpBot.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.wait import WebDriverWait
import time
from datetime import datetime



driver = webdriver.Chrome()

driver.get("https://www.buchsys.ahs.tu-dortmund.de/angebote/aktueller_zeitraum/_Badminton.html")

#function to sign up for an appointment
def signIn(path): #path is the XPATH of the book button that macht the day of week
    
    #choose the offer of the same weekday next week
    bookBut = driver.find_element(By.XPATH, path)
    bookBut.click()

    #choose to sign up with account
    driver.switch_to.window(driver.window_handles[1])
    bookBut1 = driver.find_element(By.XPATH, '//*[@id="bs_form_main"]/div/div[2]/div[1]/label/div[2]/input')
    bookBut1.click()
    signUpBut = driver.find_element(By.XPATH, '//*[@id="bs_pw_anmlink"]/div[2]')
    signUpBut.click()
    
    #fill in the register form 
    eMail = driver.find_element(By.XPATH, '//*[@id="bs_pw_anm"]/div[2]/div[2]/input')
    eMail.send_keys("abc@def.com") # ENTER YOUR USERNAME HERE
    password = driver.find_element(By.XPATH, '//*[@id="bs_pw_anm"]/div[3]/div[2]/input')
    password.send_keys("123456789") # ENTER YOUR PASSWORD HERE
    submit = driver.find_element(By.XPATH, '//*[@id="bs_pw_anm"]/div[5]/div[1]/div[2]/input')
    WebDriverWait(driver, 22).until(EC.element_to_be_clickable((By.XPATH, '//*[@id="bs_pw_anm"]/div[5]/div[1]/div[2]/input')))
    submit.click()

    #final check of the infos
    checkBox = driver.find_element(By.XPATH, '//*[@id="bs_bed"]/label/input')
    checkBox.click()
    confirm = driver.find_element(By.XPATH, '//*[@id="bs_submit"]')
    confirm.click()

    #submit the register form
    booking = driver.find_element(By.XPATH, '//*[@id="bs_foot"]/div[1]/div[2]/input')
    booking.click()


match datetime.now().weekday(): # get day of the week
    case 0: #monday
        try:
            signIn('//*[@id="bs_tr2203EF9F6C"]/td[9]/input')
        except:
            driver.quit()
    case 1:
        try:
            signIn('//*[@id="bs_tr2203EF9FD2"]/td[9]/input')
        except:
            driver.quit()
    case 3:
        try:
            signIn('//*[@id="bs_tr2203EF9F79"]/td[9]/input')
        except:
            driver.quit()
    case 4:
        try:
            signIn('//*[@id="bs_tr2203EF9F67"]/td[9]/input')
        except:
            driver.quit()
    case _:
        driver.quit()

driver.quit()



